### PR TITLE
Remove use of null-conditional operator for logger

### PR DESCRIPTION
**Pull request by Kahbazi**
Assign NullLogger.Instance for logger and remove null-conditional operator

### CODE REVIEW

**Comment by benaadams** [comment]
Will this mean it will always make an interface call rather than doing an early out on a null check?

**Comment by Kahbazi** [comment]
Well I was thinking that the logger is probably not null in most scenarios,  but you are right when it is null, the null check has better performance

File: src/Hosting/Hosting/src/Internal/WebHost.cs [hunk 1]
```diff
@@ -317,7 +317,7 @@ public async Task StopAsync(CancellationToken cancellationToken = default)
             }
             _stopped = true;
 
-            _logger?.Shutdown();
+            _logger.Shutdown();
```
Comment (Tratcher) [inline src/Hosting/Hosting/src/Internal/WebHost.cs]: Revert. It's possible (though not a good idea) to call Stop without first calling Start.

File: src/Hosting/Hosting/src/Internal/WebHost.cs [hunk 2]
```diff
@@ -364,7 +364,7 @@ public async ValueTask DisposeAsync()
                 }
                 catch (Exception ex)
                 {
-                    _logger?.ServerShutdownException(ex);
+                    _logger.ServerShutdownException(ex);
```
Comment (Tratcher) [inline src/Hosting/Hosting/src/Internal/WebHost.cs]: Revert

File: src/Middleware/Rewrite/src/Internal/ApacheModRewrite/ApacheModRewriteRule.cs [hunk 3]
```diff
@@ -26,7 +26,7 @@ public virtual void ApplyRule(RewriteContext context)
 
             if (!initMatchRes.Success)
             {
-                context.Logger?.ModRewriteNotMatchedRule();
+                context.Logger.ModRewriteNotMatchedRule();
```
Comment (Tratcher) [inline src/Middleware/Rewrite/src/Internal/ApacheModRewrite/ApacheModRewriteRule.cs]: @jkotalik why were these nullable?

Comment (jkotalik) [reply src/Middleware/Rewrite/src/Internal/ApacheModRewrite/ApacheModRewriteRule.cs]: Probably just a mistake.

File: src/Hosting/Hosting/src/Internal/WebHost.cs (hunk 1 above)
Comment (Kahbazi) [reply src/Hosting/Hosting/src/Internal/WebHost.cs]: What if I assign `NullLogger` to `_logger` in constructor?

Comment (Tratcher) [reply src/Hosting/Hosting/src/Internal/WebHost.cs]: That could work.

**Review by Tratcher — APPROVED** [review]

**Review by halter73 — APPROVED** [review]

**Comment by halter73** [comment]
@sebastienros If we merge this, we might want to watch the benchmarks. Normally I wouldn't worry about virtual dispatch, but logging can be pretty frequent.

**Comment by Tratcher** [comment]
@sebastienros  @halter73 most of these were only ever null in tests so product benchmarks shouldn't be affected.

**Comment by Tratcher** [comment]
Thanks

**Pull request merged**